File: bottomsheetmenu/src/main/java/com/strekha/bottomsheetmenu/MenuAdapter.java
package com.strekha.bottomsheetmenu;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.List;

final class MenuAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TITLE = 101;
    private static final int ELEMENT = 102;
    private static final int SEPARATOR = 103;

    private final List<Item> items;
    private final BottomSheetMenu.OnBottomMenuListener listener;
    private final int itemType;

    MenuAdapter(@NonNull List<Item> items,
                @NonNull BottomSheetMenu.OnBottomMenuListener listener,
                int itemType) {
        this.items = items;
        this.listener = listener;
        this.itemType = itemType;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        switch (viewType) {
            case TITLE:
                return new Holder.Title(inflater, parent, itemType);
            case ELEMENT:
                return new Holder.Element(inflater, parent, listener, itemType);
            case SEPARATOR:
                return new Holder.Separator(inflater, parent, itemType);
            default:
                throw new IllegalArgumentException("Unknown viewType!");
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Item item = items.get(position);
        switch (getItemViewType(position)) {
            case TITLE:
                ((Holder.Title) holder).bind((Item.Title) item);
                break;
            case ELEMENT:
                ((Holder.Element) holder).bind(((Item.Element) item));
                break;
            case SEPARATOR:
                break;
            default:
                throw new IllegalArgumentException("Unknown viewType!");
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public int getItemViewType(int position) {
        Item item = items.get(position);
        if (item instanceof Item.Title) {
            return TITLE;
        } else if (item instanceof Item.Element) {
            return ELEMENT;
        } else if (item instanceof Item.Separator) {
            return SEPARATOR;
        }
        throw new IllegalArgumentException("Unknown viewType!");
    }

    @NonNull
    Item getItem(int position) {
        return items.get(position);
    }
}

File: bottomsheetmenu/src/main/java/com/strekha/bottomsheetmenu/MenuAdapter.kt
package com.strekha.bottomsheetmenu

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.ViewGroup

internal class MenuAdapter(
        private val items: List<Item>,
        private val itemType: Int,
        private val listener: (MenuItem) -> Unit
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return when (viewType) {
            TITLE -> Holder.Title(inflater, parent, itemType)
            ELEMENT -> Holder.Element(inflater, parent, listener, itemType)
            SEPARATOR -> Holder.Separator(inflater, parent, itemType)
            else -> throw IllegalArgumentException("Unknown viewType!")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = items[position]
        when (getItemViewType(position)) {
            TITLE -> (holder as Holder.Title).bind(item as Item.Title)
            ELEMENT -> (holder as Holder.Element).bind(item as Item.Element)
            SEPARATOR -> { /* ignored */}
            else -> throw IllegalArgumentException("Unknown viewType!")
        }
    }

    override fun getItemCount() = items.size

    override fun getItemViewType(position: Int): Int {
        val item = items[position]
        return when (item) {
            is Item.Title -> TITLE
            is Item.Element -> ELEMENT
            is Item.Separator -> SEPARATOR
        }
    }

    fun getItem(position: Int): Item {
        return items[position]
    }

    companion object {

        private const val TITLE     = 101
        private const val ELEMENT   = 102
        private const val SEPARATOR = 103
    }
}
